src/python/pants/option/config.py

```diff
@@ -230,10 +244,173 @@ def get_value(self, section: str, option: str) -> Optional[str]:
   def options(self, section: str) -> List[str]:
     return self.parser.options(section)
 
+  @property
   def defaults(self) -> Mapping[str, str]:
     return self.parser.defaults()
 
 
+_TomlPrimitve = Union[bool, int, float, str]
+_TomlValue = Union[_TomlPrimitve, List[_TomlPrimitve]]
+
+
+@dataclass(frozen=True)
+class _TomlValues(_ConfigValues):
+  values: Dict[str, Any]
+
+  @staticmethod
+  def _section_explicitly_defined(section_values: Dict) -> bool:
+    """Because of the way TOML stores the config as a nested dictionary, it naively appears that
+    certain sections are defined when really only a subscope is defined.
+
+    For example, the user may have specified `cache.java` but not `cache`.
+    """
+    at_least_one_option_defined = any(
+      not isinstance(section_value, dict) for section_value in section_values.values()
+    )
+    blank_section = len(section_values.values()) == 0
+    return at_least_one_option_defined or blank_section
+
+  def _find_section(self, section: str) -> Optional[Dict]:
+    def recurse(mapping: Dict, *, remaining_sections: List[str]) -> Optional[Dict]:
+      if not remaining_sections:
+        return None
+      current_section = remaining_sections[0]
+      if current_section not in mapping:
+        return None
+      section_values = mapping[current_section]
+      if len(remaining_sections) > 1:
+        return recurse(section_values, remaining_sections=remaining_sections[1:])
+      if not self._section_explicitly_defined(section_values):
+        return None
+      return cast(Dict, section_values)
+
+    return recurse(mapping=self.values, remaining_sections=section.split("."))
+
+  def _possibly_interpolate_value(self, raw_value: str) -> str:
+    """For any values with %(foo)s, substitute it with the corresponding default val."""
+    def format_str(value: str) -> str:
+      # Because dictionaries use the symbols `{}`, we must proactively escape the symbols so that
+      # .format() does not try to improperly interpolate.
+      escaped_str = value.replace("{", "{{").replace("}", "}}")
+      new_style_format_str = re.sub(
+        pattern=r"%\((?P<interpolated>[a-zA-Z_0-9]*)\)s",
+        repl=r"{\g<interpolated>}",
```

Comment:
Generally, it's best practice to use `re.compile`. But it looks like that's not necessary in modern Python due to automatic caching? https://docs.python.org/3/library/re.html#re.compile

Lmk if I should use `re.compile` and any tips for where to put that - I'm not sure if it needs to be declared as a module constant, rather than a variable belonging to the method.

Reply:
If python3 automatically caches regexes, and these regexes are only used once on options parsing and then not needed for the rest of the run, then it's probably best *not* to `re.compile` at the module level, so I think this is good as-is.